crates/shared/src/token_info.rs

```diff
@@ -1,173 +1,239 @@
-                )
-            })
-            .collect::<Vec<_>>();
-
-        batch.execute_all(MAX_BATCH_SIZE).await;
-        let mut resolved_futures = Vec::with_capacity(futures.len());
-        for (decimals, symbol) in futures {
-            resolved_futures.push((decimals.await, symbol.await));
+    async fn get_token_info(&self, address: H160) -> Result<TokenInfo, Error> {
+        let info = self.fetch_token(address).await;
+        if let Err(err) = &info {
+            tracing::debug!(?err, token = ?address, "failed to fetch token info");
         }
-        addresses
-            .iter()
-            .zip(resolved_futures)
-            .map(|(address, (decimals, symbol))| {
-                if decimals.is_err() {
-                    tracing::trace!("Failed to fetch token info for token {}", address);
-                }
-                (
-                    *address,
-                    TokenInfo {
-                        decimals: decimals.ok(),
-                        symbol: symbol.ok(),
-                    },
-                )
-            })
-            .collect()
+
+        info
+    }
+
+    async fn get_token_infos(&self, addresses: &[H160]) -> HashMap<H160, TokenInfo> {
+        futures::future::join_all(addresses.iter().copied().map(|address| async move {
+            let info = self.fetch_token(address).await;
+            if let Err(err) = &info {
+                tracing::debug!(?err, token = ?address, "failed to fetch token info");
+            }
+
+            (address, info.unwrap_or_default())
+        }))
+        .await
+        .into_iter()
+        .collect()
     }
 }
 
+type SharedTokenInfo = Shared<BoxFuture<'static, Result<TokenInfo, Error>>>;
+
 pub struct CachedTokenInfoFetcher {
-    inner: Box<dyn TokenInfoFetching>,
-    cache: Arc<Mutex<HashMap<H160, TokenInfo>>>,
+    inner: Arc<dyn TokenInfoFetching>,
+    cache: Arc<Mutex<HashMap<H160, SharedTokenInfo>>>,
 }
 
 impl CachedTokenInfoFetcher {
-    pub fn new(inner: Box<dyn TokenInfoFetching>) -> Self {
+    pub fn new(inner: Arc<dyn TokenInfoFetching>) -> Self {
         Self {
             inner,
             cache: Arc::new(Mutex::new(HashMap::new())),
         }
     }
 }
 
+impl CachedTokenInfoFetcher {
+    async fn fetch_token(&self, address: H160) -> Result<TokenInfo, Error> {
+        let fetch = {
+            let mut cache = self.cache.lock().unwrap();
+            cache
+                .entry(address)
+                .or_insert({
+                    let inner = self.inner.clone();
+                    async move { inner.get_token_info(address).await }
+                        .boxed()
+                        .shared()
+                })
+                .clone()
+        };
+
+        let info = fetch.await;
+        if info.is_err() {
+            let mut cache = self.cache.lock().unwrap();
+            if let Some(Err(_)) = cache.get(&address).and_then(|fetch| fetch.peek()) {
```

Comment:
What is this condition checking? Can't we just uncoditionally remove the future in case of error?

Reply:
It is checking that the shared future stored in the map is resolved to an error (with `Shared::peek`).

Imagine 3 concurrent requests `Rn` where:
1. `R1` starts a shared fetch of some token information
2. `R2` uses the same shared fetch
3. The shared future resolves to an error, `R1` removes the shared future from the cache
4. `R3` starts a new shared fetch of token information for the same address, the cache entry was removed in 3, so it creates a new one
5. `R2` gets scheduled and removes the new shared future that was created by `R3`.

This isn't extremely contrived, as there is an `await` point before clearing the cache, so if `R2` doesn't get polled for some time, this would happen. That being said, I don't think this will happen regularly, and at worse we will be slightly sub-optimal and duplicate some queries.

Another approach would be to `peek` at the shared future when we get the first cache lock and replace it use the shared future in case it is an error. The downside being that we can hold some useless `Error(string)`s in the cache that won't ever get used.